### ISSUE TITLE
TransFormer 中的 层归一化问题

### ISSUE THREAD

**Issue by dongbo910220** [Hoper-J/AI-Guide-and-Demos-zh_CN#10]
您好 在 TransFormer 中提到了层归一化问题， 我理解mean = x.mean(dim=(-2, -1), keepdim=True) 更像是论文表达的原意， 包括李沐的视频中也画了一个切面， 但实际上您这里用的是  mean = x.mean(dim=-1, keepdim=True) 也就是仅仅对某个词进行了归一化， 是不是有点不符合论文原意呢？

下面是您的实现

class LayerNorm(nn.Module):
    def __init__(self, feature_size, epsilon=1e-9):
        """
        层归一化，用于对最后一个维度进行归一化。
        
        参数:
            feature_size: 输入特征的维度大小，即归一化的特征维度。
            epsilon: 防止除零的小常数。
        """
        super(LayerNorm, self).__init__()
        self.gamma = nn.Parameter(torch.ones(feature_size))  # 可学习缩放参数，初始值为 1
        self.beta = nn.Parameter(torch.zeros(feature_size))  # 可学习偏移参数，初始值为 0
        self.epsilon = epsilon

    def forward(self, x):
        mean = x.mean(dim=-1, keepdim=True)
        std = x.std(dim=-1, keepdim=True)
        return self.gamma * (x - mean) / (std + self.epsilon) + self.beta

**Comment by Hoper-J** [comment]
很好的问题，LayerNorm 本身的命名容易会让人误解，实际上"Layer"在这里并非是指层这个几何意义上的二维，而是神经网络中的某一层（"层"包含了多个神经元，每个神经元对应一个特征维度）。

更准确地说，LayerNorm 在 Transformer 的使用中仅仅是对同一层内的神经元激活值进行归一化。假设输入张量形状为[batch_size, seq_len, hidden_size]，LayerNorm 就是作用于最后一个维度（hidden_size），也正是“仅仅对某个词进行了归一化”，你的感觉是对的 :) 

下面有一些相关的链接：
1. [Transformer官方代码layer_norm部分实现](https://github.com/tensorflow/tensor2tensor/blob/bafdc1b67730430d38d6ab802cbd51f9d053ba2e/tensor2tensor/layers/common_layers.py#L715)，其中的注释提及：“Layer normalize the tensor x, averaging over the last dimension.”
2. [Why do transformers use layer norm instead of batch norm?](https://stats.stackexchange.com/questions/474440/why-do-transformers-use-layer-norm-instead-of-batch-norm)：其中的讨论或许会给你带来思路。
3.  [Layernorm is confusingly named?](https://www.reddit.com/r/MachineLearning/comments/1g7am36/d_layernorm_is_confusingly_named/)
4. [What is Layer Normalization?](https://www.youtube.com/watch?v=2V3Uduw1zwQ)：这个视频的例子不太直观，但是中间的讲解可以看下（1:30）

希望上述的回答可以解决你的一些疑惑。

题外话：了解线性层的参数维度和神经元可能会更清晰，线性层本身在 shape 的角度上看也只是对最后一个维度进行了变化。

**Comment by dongbo910220** [comment]
非常感谢! 解决了我的困惑
我发现其实之前对layer 的理解受到了李沐老师的那个画切面图的影响， 现在再把您提供的资料都看到之后， 我明白了您提到的”线性层本身在 shape 的角度上看也只是对最后一个维度进行了变化。” 
此外还有个问题想请教下， 您是之前也有这方面的困惑么？ 我看您很快就找到了相关的资料， 而我可能要花比较多的时间， 不知您是如何在短时间内找到比较多的很有价值的资料？
再次感谢！

**Comment by Hoper-J** [comment]
这方面的困惑的确在之前出现过，所以能够 Get 到。相关的资料是临时找的，当时的思路大致如下：

看到问题的瞬间先换位思考，假定文章代码实现是错误的，即“Transformer 中的 LayerNorm 应该处理两个维度”，但该观点缺少一个关键论据：Transformer 的实际实现。
1. 所以第一步：访问[官方代码仓库](https://github.com/tensorflow/tensor2tensor/)搜索 "Layer" 查看相关代码，[对应实现](https://github.com/tensorflow/tensor2tensor/blob/bafdc1b67730430d38d6ab802cbd51f9d053ba2e/tensor2tensor/layers/common_layers.py#L715)表明就是使用最后一个维度。
2. 最初的论点“Transformer 中的 LayerNorm 应该处理两个维度”被推翻，衍生出可能的两个疑惑：
    - Transformer 的 LayerNorm 使用是否正确？（即便是官方，也应保持质疑）
        - 搜索（实际为英文）：Transformer LayerNorm, 为什么是 LayerNorm, 什么是 LayerNorm。
    - LayerNorm 在不同语境下是否有不同定义？（这个问题是“常驻嘉宾”，起源于对 GPT 系列论文的阅读，其中对少样本等说法进行了重定义，在当时给我造成了概念上的混淆，所以后面在质疑论文本身的时候我都会先考虑这一点，如果是重定义，就代表第一个疑惑被证实：并非正确使用，因为重新进行了定义）
        - LayerNorm 如果存在不同定义，那一定是因为 “Layer” 这个命名本身，从而有以下搜索路径（实际为英文）：LayerNorm 一维，为什么不是二维？LayerNorm 中的 Layer 指的是什么？

相关资料完全基于上述关键词查找。

---

其实说到底，就是抓住自己真正的疑惑，然后挖掘背后的理论和代码实现。

如果对 LayerNorm 有疑惑，那么就去探索 LayerNorm 的相关源代码&最初论文&视频&数学原理。

如果对 Layer 有疑惑，那就不需要关注 Norm 的实际数学公式，而是探索：为什么是 Layer？LayerNorm 中的 Layer 对象是什么？

相信在学习的道路上一定有人和自己有着同样的疑惑，发出同样的疑问，且已经被热心的同学解答，运用好搜索引擎。

（如果没有，那么或许这正是一个新的点子等待发掘）

**Comment by dongbo910220** [comment]
感谢感谢！可能有了大模型之后， 自己在搜索上就做的不够了，加上很多时候自己可能搜索方法不对，所以自己会没有充分利用好， 非常好的提醒！

下面是针对您提到的：

​	1.首先换位思考部分其实我就不一定能做到， 有时候发现不对了， 只会找论文反复琢磨， 而忽略了	Transformer本身实现是如何实现。后来我自己又查找确认了一遍，理解也更清晰了

```
transformer_encoder()  
    → layer_postprocess()  
        → layer_prepostprocess()  
            → apply_norm(norm_type="layer")  
                → layer_norm()  
                    → layer_norm_vars()  # 获取γ和β参数  
                    → layer_norm_compute()  # 实际计算
```

 	2. 针对您的回答给了我很大启发， 我也沿着您的思路都搜索了一遍， 也找到了对应的资料。但还有几个问题想问下？
     	1. 为什么中文的我看您都没有引用？ 是因为大多质量不一定能得到保证么？ 我之前查找效果不好，是否也因为主要搜索了中文资料？
     	2. 如果搜索中， 总是有很多文章的， 比如有篇博客的名字叫  LayerNorm == InstanceNorm?  "如果遇到这种情况，您会看这篇文章吗？ （我觉得自己很多时候搜索出来的东西一多， 看几篇就乱了）
     	3. 这个问题和上一个有些类似， 有时候看多了就容易钻进去了， 忘了自己要干嘛了， 比如我看了很多“ 为什么是 LayerNorm, 什么是 LayerNorm” 之后， 就忘了衍生出来的第二个疑惑 “LayerNorm 一维，为什么不是二维？LayerNorm 中的 Layer 指的是什么？”， 想问下您也会有类似的状况么？ 其实我觉得您思路非常清晰， 应该不太会有类似的情况。

最后祝生活愉快， 工作顺利， 发掘很多新点子~

**Comment by Hoper-J** [comment]
1. 很高兴能有实际的帮助，中文也是会搜索和引用的，当时粗略地看了下对应回答和视频讲解的质量发现不够相关和直观就没引用。查找效果不好很可能是因为主要搜寻了中文资料，一般而言，我会两个都进行搜寻。中文的基础理论挺全的，但所延伸出的问答比较少，知乎基本是国内唯一有价值的问答平台，而国外类似的平台和网站有很多，同时因为论文本身是英文的，英文搜索效果会好非常多，单以我过去的搜索结果来看，英文所提供的相关信息>>中文，英文没有解决的问题基本上中文也找不到解决方式。
对于新的知识，基础数学&公式原理部分我通常先会查看中文文章（因为中文作为母语本身就有着理解优势），出现疑惑也会先中文搜索，如果第一眼没有相关文章才会转为英文。

2. InstanceNorm 这篇文章会进行查看，但不会过度深究，因为单 token 仔细一想确实跟 Instance 想法相似。
在精准的关键词搜索下靠前的文章一定会有一些语义相关的信息，或许能够帮助自己触类旁通。

3. 当然是会有的:) 同学不要妄自菲薄，保持探索精神，这非常棒，曾经我也是一个喜欢钻“[rabbit hole](https://en.wikipedia.org/wiki/Wiki_rabbit_hole)”的人（甚至还因为1+1=2去读了陶哲轩的实分析而推迟了主线）。
需要做好权衡，弄清楚细节很有意义的，可或许有时候你并不需要了解那么深，更多的将“探索”作为自己的课余爱好（每天为它分配一定的时间），别因为它耽误了今天应该推进的学习主线。

---

提到“换位思考”主要是因为要思考引用哪些文章可能对理解问题有帮助，实际上在正常的学习过程中用的很少，因为我一般也处于提问和质疑的状态。

另外，大模型的输出只能当作旁证，对于新的知识，我们并不知道大模型的输出是否正确，但过程中的理论依据都是有现成资料可以查验的，通过查验去学习，通过学习再质疑，只有在它给出的理论推理能够说服你的时候才真正的相信。
【质疑->修正->直到大模型的输出有理有据】这一循环与大模型交互的过程实际上是很好的学习方法，一种双向的知识奔赴。

谢谢同学的祝福，也祝你学业有成，科研顺利～

（Issue 就暂时关闭啦，如果有新的想法可以开启新的 [Issue](https://github.com/Hoper-J/AI-Guide-and-Demos-zh_CN/issues/new/choose) 或者 [Discussion](https://github.com/Hoper-J/AI-Guide-and-Demos-zh_CN/discussions)）。
